### ISSUE TITLE
Django - tarea de conteo

### ISSUE THREAD

**Issue by estephany-up** [estephany-up/Incentivos_ie#2]
Hola chicos!
Como les comenté el día miércoles, encontré un código de la tarea de contar ceros. Lo que tengo entendido es que usan todo lo que hay en "class Task(djmodels.Model)" para crear la tabla y me parece que también para verificar las respuestas. De esta manera, utilizan Task en:
def get_or_create_task(self):
        unfinished_tasks = self.tasks.filter(answer__isnull=True)
        if unfinished_tasks.exists():
            return unfinished_tasks.first()
        else:
            task = Task.create(self, self.session.vars['task_fun'], **self.session.vars['task_params'])
            task.save()
            return task
Y esta función la llaman en la página donde irá la tarea, en mi caso es en Tarea_conteo. Para la esta tarea se usa html: conteo, tarea_conteo y js_wp; así como py: models y ret_fun.
Lo que no tengo muy claro es como utilizar djmodels, aunque me parece que en esta versión de otree no se usa. No sabría como reformular esa parte de class Task.

**Comment by Sergio2405** [comment]
Buenos días, en todo caso no puedan reformular el codigo para que funcione en otree, les recomiendo que traten de ustedes mismas escribir su propio codigo. A continuación les adjunto un link en donde pueden encontrar mas informacion sobre como realizar tablas en HTML. https://www.tablesgenerator.com/html_tables#
@estephany-up

**Comment by estephany-up** [comment]
Justo estaba intentando crearlo desde cero y me he quedado estancada. Mi problema, en si, es cómo llenar la tabla porque no lo puedo hacer simplemente poniendo 1 o 0, el valor que va dentro de la celda debe ser elegido aleatoriamente. Para ello, estoy haciendo la tabla de la forma larga y luego cuando tenga todo, reducirlo con un "for". Lo que he estado intentando hacer es crear variables para cada celda (para empezar sólo pensaba llenar 5 celdas) y que el valor de cada variable sea elegido aleatoriamente de una lista: ['0','1']. Después, llamo a la variable en el html. Por alguna razón me sale error lo he intentado hacer desde models y también desde pages. 
Para el caso de models he usado esto:
class Task:
    a_1 = models.StringField()
    a_2 = models.StringField()
    a_3 = models.StringField()
    a_4 = models.StringField()
    a_5 = models.StringField()
    
    def get_body(self, **kwargs):
        selection_set = kwargs.get('selection_set', [0, 1])
        for i in range (1,5):
            a_i = str(random.choice(selection_set))
**he definido un class Task porque el valor no es único para el participante y pensé que en players no sería apropiado ponerlo. En realidad, la tabla cambiará cada vez que el participante agregue una respuesta y se encuentre dentro del tiempo establecido. 

Para el caso de pages he usado esto:
def vars_for_template(self):
        for i in range (1,5):
            a_i = random.choice([Constants.selection_set]) 

Tengo el presentimiento de que he definido algo mal o en el lugar incorrecto. 
**PD: en Tarea_conteo.html pongo la variable de la siguiente manera --> {{a_1}}

**Comment by Sergio2405** [comment]
Buenos días, esta bien la lógica que estas siguiendo. Sin embargo, si no me equivoco, el problema puede venir en la parte de "pages" en el método "vars for template", ya que este ultimo debe retornar un diccionario que contenga los valores de los "a_i". A continuación les dejo un ejemplo: 
```
def vars_for_template(self):
    value = Constants.endowment
    return dict( 
             x = value
           )
ese x es el que va en los templates :)
```
Por los demás que mencionas no creo que haya algún problema 
           @estephany-up

**Comment by estephany-up** [comment]
Muchas gracias! Definí las variables en pages y luego las llame en el template respectivo. Al final, pude crear la tabla llena de unos y ceros. Y ahora viene la parte que me parece que es más difícil porque ahora la tabla tiene que cambiar de valor cada vez que el participante presione el botón de siguiente, pero todo esto debe de ocurrir en un mismo template ¿no?
Lo que tenía pensado es, tal vez, poner una función donde cada vez que presione siguiente y se encuentre dentro del tiempo establecido, cambie los valores dentro de la tabla. Sin embargo, no sé que tanto puedan cambiar en mi caso porque creo que las variables que definí son fijas. No sé si hay otra forma de plantear la actualización de los valores que tiene cada celda de la tabla.

**Comment by Sergio2405** [comment]
Buenazo! si claro una funcion donde cambien los valores de la tabla estaría bien. Con respecto a lo que se mantienen fijas, puede ser debido a que estas creando las variables "a_i" de frente en "pages". Una posible solución seria crear estas variables en "models" dentro de una funcion que le asigne valores aleatorios a todos los "a_i".
@estephany-up

**Comment by estephany-up** [comment]
Claro, sería algo parecido a lo que quería hacer en models antes. He definido esta función:
def func(n):
    lst=[]
    for i in range (1,n):
        a_i = str(random.choice([0,1]))
        lst.append(a_i)
    return (lst)

class Task:
    lt_1 = func(10)
    lt_2 = func(10)
    lt_3 = func(10)
    lt_4 = func(10)
    lt_5 = func(10)
    lt_6 = func(10)
    lt_7 = func(10)
    lt_8 = func(10)
    lt_9 = func(10)
    lt_10 = func(10)
Estoy trabajando con listas, lo he probado y si me corre, pero sólo cuando la lista la pongo en el vars for template. Entonces, ¿sería lo mismo que hacerlo en pages? O en este caso, ¿ya puedo definir otra función que cambie las variables? Una función que contenga la función func que ya definí.

**Comment by Sergio2405** [comment]
Si corre, perfecto. Las siguientes funciones que desees crear seria mejor que las crees en el models.

**Comment by estephany-up** [comment]
Hola! He estado tratando de crear la función que tenía planeada (donde cada vez que presione siguiente y se encuentre dentro del tiempo establecido, cambie los valores dentro de la tabla), pero no lo he logrado porque no he encontrado ningún "nombre de alguna variable u objeto" como para tomarlo y que dependiendo de su estado haga la actualización de los valores de la tabla. En mi caso, tenía planeado tomar el botón de siguiente; sin embargo, no hallo una forma de llamarlo en models, donde supuestamente planeo llamar a la función. Y sólo se que con el botón siguiente se puede pasar a la siguiente ronda o página. No sé si alguno de ustedes tiene una noción de cómo lo podría plantear.

**Comment by Sergio2405** [comment]
Una forma en la que lo pondrían plantear seria crear varias rondas en donde para cada una de estas halla una tabla diferente. Es decir para la ronda 1 esta la tabla 1, para la ronda 2 otra tabla y así. Estas tablas serian iguales para todos los jugadores. Y el tiempo por cada ronda seria de aprox 20 segundos y podrían hacer 10 rondas y esto les daría en total los 3 minutos que tenían pensados. Asimismo, para que sea simple y se pueden efectuar las 2 tareas (contar y el de escribir palabras) , separen cada tarea en 2 apps distintas (2 repositorios). Las aplicaciones se pueden juntar tranquilamente al momento de correr el experimento.

**Comment by estephany-up** [comment]
Gracias, creo que ese modo está mejor.
Ya cree otra app en donde está mi app de incentivos con el comando otree startapp NOMBRE. También, ya pasé todas las variables y funciones necesarias para la tarea de conteo. Pero, a la hora de correr el código me sale este mensaje:

![mensaje de error](https://user-images.githubusercontent.com/80181385/126011224-34160861-e3b2-403f-ae2a-9718c9cb1d20.PNG)

No sé a qué se pueda deber ese error. Lo que he encontrado es que hay una carpeta que se llama "temp_migrations" y hay varias carpetas y una de ellas se llama incentivos que es la primera app que cree, pero no está la app que acabo de crear. Tal vez tenga algo que ver con ese error.

**Comment by Sergio2405** [comment]
mm me parece raro ese error. Asegúrense que los 3 archivos, las 2 app y el settings.py, se encuentren en la misma carpeta. Asimismo, en el settings.py debe haber algo similar a esto: 
```
SESSION_CONFIGS = [
    {
        'name': 'incentivos',
        'display_name': "Incentivos",
        'num_demo_participants': 1,
        'app_sequence': ['incentivos', 'la otra app']
   }
```
En todo caso les recomiendo revisar la parte de 'Apps & Rounds' de la documentación de otree
@estephany-up

**Comment by Sweetdreamstome** [comment]
También checkeen su models.py o pages.py y fíjense que nombres le han puesto a los 'nombre.html' que hay en esos dos .py, a veces hay que hacer cambios a los nombres de los templates al cambiar de apps.

**Comment by estephany-up** [comment]
Muchas gracias a ambos! Ya pude encontrar el error. Lo más probable es que más tarde este compartiendo el segundo repositorio.
